File: src/main/java/com/youngjun/auth/core/api/security/filter/ApiAuthenticationEntryPoint.java
package com.youngjun.auth.core.api.security.filter;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.youngjun.auth.core.api.support.error.AuthErrorType;
import com.youngjun.auth.core.api.support.response.AuthResponse;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.MediaType;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.AuthenticationEntryPoint;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.nio.charset.StandardCharsets;

@Component
public class ApiAuthenticationEntryPoint implements AuthenticationEntryPoint {

    private static final Logger log = LoggerFactory.getLogger(ApiAuthenticationEntryPoint.class);

    private final ObjectMapper objectMapper;

    public ApiAuthenticationEntryPoint(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    @Override
    public void commence(HttpServletRequest request, HttpServletResponse response, AuthenticationException ex)
            throws IOException {
        AuthErrorType authErrorType = resolve(ex);
        log(ex, authErrorType);
        write(response, authErrorType);
    }

    private AuthErrorType resolve(AuthenticationException ex) {
        AuthErrorType authErrorType;
        if (ex instanceof BadTokenException) {
            authErrorType = AuthErrorType.TOKEN_INVALID_ERROR;
        } else if (ex instanceof BadCredentialsException) {
            authErrorType = AuthErrorType.AUTH_BAD_CREDENTIALS_ERROR;
        } else {
            authErrorType = AuthErrorType.UNAUTHORIZED_ERROR;
        }
        return authErrorType;
    }

    private void log(AuthenticationException ex, AuthErrorType authErrorType) {
        switch (authErrorType.getLogLevel()) {
            case ERROR -> log.error("AuthException : {}", ex.getMessage(), ex);
            case WARN -> log.warn("AuthException : {}", ex.getMessage(), ex);
            default -> log.info("AuthException : {}", ex.getMessage(), ex);
        }
    }

    private void write(HttpServletResponse response, AuthErrorType authErrorType) throws IOException {
        response.setStatus(authErrorType.getStatus().value());
        response.setContentType(MediaType.APPLICATION_JSON_VALUE);
        response.setCharacterEncoding(StandardCharsets.UTF_8.name());
        response.getWriter().write(objectMapper.writeValueAsString(AuthResponse.error(authErrorType)));
    }

}

File: src/main/kotlin/com/youngjun/auth/api/security/ApiAuthenticationEntryPoint.kt
package com.youngjun.auth.api.security

import com.fasterxml.jackson.databind.ObjectMapper
import com.youngjun.auth.api.support.error.ErrorType
import com.youngjun.auth.api.support.response.AuthResponse
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.boot.logging.LogLevel
import org.springframework.http.MediaType
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.AuthenticationEntryPoint
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets

@Component
class ApiAuthenticationEntryPoint(
    private val objectMapper: ObjectMapper,
) : AuthenticationEntryPoint {
    private val log: Logger = LoggerFactory.getLogger(javaClass)

    override fun commence(
        request: HttpServletRequest,
        response: HttpServletResponse,
        ex: AuthenticationException,
    ) {
        val errorType = resolve(ex)
        log(ex, errorType)
        write(response, errorType)
    }

    private fun resolve(ex: AuthenticationException) =
        when (ex) {
            is BadTokenException -> {
                ErrorType.TOKEN_INVALID_ERROR
            }

            is BadCredentialsException -> {
                ErrorType.AUTH_BAD_CREDENTIALS_ERROR
            }

            else -> {
                ErrorType.UNAUTHORIZED_ERROR
            }
        }

    private fun log(
        ex: AuthenticationException,
        errorType: ErrorType,
    ) {
        when (errorType.logLevel) {
            LogLevel.ERROR -> log.error("AuthException : {}", ex.message, ex)
            LogLevel.WARN -> log.warn("AuthException : {}", ex.message, ex)
            else -> log.info("AuthException : {}", ex.message, ex)
        }
    }

    private fun write(
        response: HttpServletResponse,
        errorType: ErrorType,
    ) {
        response.status = errorType.status.value()
        response.contentType = MediaType.APPLICATION_JSON_VALUE
        response.characterEncoding = StandardCharsets.UTF_8.name()
        response.writer.write(objectMapper.writeValueAsString(AuthResponse.error(errorType)))
    }
}
